byterun/memory.c

```diff
@@ -586,7 +586,7 @@ CAMLexport CAMLweakdef void caml_initialize (value *fp, value val)
 {
   CAMLassert(Is_in_heap(fp));
   *fp = val;
-  if (Is_block (val) && Is_young (val)) {
+  if (!Is_young((value)fp) && Is_block (val) && Is_young (val)) {
```

Comment:
I'm unsure what this change is for.

Currently, `caml_initialize` only works on the major heap, and it is an error to use it on minor heap objects. You seem to be relaxing this restriction with the new check (which I think is a good idea, multicore already works that way).

However, you're careful to not use the newly relaxed caml_initialize (inlining the minor heap check in cmmgen.ml), and you haven't changed the `Is_in_heap` assertion to `Is_in_heap_or_young`, so I don't know what your intent is.